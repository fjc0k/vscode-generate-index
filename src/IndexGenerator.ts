import * as changeCase from 'change-case'
import globby, { GlobbyOptions } from 'globby'
import { castArray, isRegExp, noop } from 'vtils'
import {
  CodeGenerator,
  CodeGeneratorRe,
  Marker,
  ParsedPath,
  Pattern,
} from './types'
import { dirname, join, parse, relative } from 'path'
import { readFile, statSync } from 'fs-extra'

export interface IndexGeneratorOptions {
  filePath: string
  fileContent: string
  onGenerate: (payload: { code: string; marker: Marker }) => any
  onWarning: (msg: string) => any
}

export class IndexGenerator {
  constructor(private options: IndexGeneratorOptions) {}

  public async generate(): Promise<false | void> {
    const { filePath, fileContent, onWarning, onGenerate } = this.options
    const fileDir = dirname(filePath)
    const markers = IndexGenerator.findMarkers(fileContent, onWarning)
    if (markers.length) {
      for (const marker of markers) {
        if (!marker.isRe) {
          const paths = await globby(marker.patterns, {
            dot: true,
            onlyFiles: false,
            // TODO: fix https://github.com/sindresorhus/globby/issues/133
            gitignore: false,
            ...(marker.globbyOptions || {}),
            cwd: fileDir,
            absolute: true,
          })
          paths.sort(
            new Intl.Collator(['en', 'co'], {
              sensitivity: 'base',
              numeric: true,
            }).compare,
          )
          const codes = paths
            .filter(path => path !== filePath)
            .map((path, index, paths) => {
              const pp = parse(path)
              const parsedPath: ParsedPath = {
                path: IndexGenerator.getRelativePath(
                  fileDir,
                  join(pp.dir, pp.name),
                ),
                name: pp.name,
                ext: pp.ext,
              }
              const code = marker.codeGenerator(parsedPath, changeCase, {
                get total() {
                  return paths.length
                },
                get index() {
                  return index
                },
                get first() {
                  return index === 0
                },
                get last() {
                  return index === paths.length - 1
                },
                get isFirst() {
                  return index === 0
                },
                get isLast() {
                  return index === paths.length - 1
                },
                get isDir() {
                  return statSync(path).isDirectory()
                },
                get isFile() {
                  return statSync(path).isFile()
                },
              })
              return marker.indent + code
            })
          await onGenerate({
            marker: marker,
            code: `${marker.start === 0 ? '' : '\n'}${codes.join('\n')}\n`,
          })
        } else {
          const sourceContent = await readFile(
            join(fileDir, marker.fileRe),
            'utf-8',
          )
          const region = sourceContent.match(marker.regionRe)?.[0] || ''
          // @ts-ignore
          const matches = [...(region.matchAll(marker.matchRe) || [])]
          const codes = matches.map((match, index, { length }) => {
            const code = marker.codeGeneratorRe(match, changeCase, {
              get total() {
                return length
              },
              get index() {
                return index
              },
              get isFirst() {
                return index === 0
              },
              get isLast() {
                return index === length - 1
              },
            })
            return marker.indent + code
          })
          await onGenerate({
            marker: marker,
            code: `${marker.start === 0 ? '' : '\n'}${codes.join('\n')}\n`,
          })
        }
      }
    } else {
      onWarning('No @index maker found.')
      return false
    }
  }

  public static findMarkers(
    text: string,
    onInvalid?: (msg: string) => any,
  ): Marker[] {
    const markers: Marker[] = []

    const textSize = text.length
    let startFrom = 0
    while (startFrom < textSize) {
      const part = text.substr(startFrom)
      const startMatch = part.match(
        /([^\r\n]*)@index\(([^\r\n]+)\)[^\r\n]*(?=[\r\n]|$)/,
      )
      if (!startMatch) break
      const endMatch = part.match(/[^\r\n]*@endindex/)
      const start = startFrom + startMatch.index! + startMatch[0].length
      const end = !endMatch ? text.length : startFrom + endMatch.index!
      const indent = startMatch[1].match(/^\s*/)![0]
      let patterns: Pattern[] = []
      let codeGenerator: CodeGenerator = noop
      let globbyOptions: GlobbyOptions = {}
      let isRe = false
      let fileRe = ''
      let codeGeneratorRe: CodeGeneratorRe = noop
      let regionRe = /1/
      let matchRe = /1/
      // eslint-disable-next-line no-inner-declarations
      function setParams(...args: any[]) {
        if (isRegExp(args[1])) {
          // eslint-disable-next-line prefer-const
          let [_file, _regionRe, _matchRe, _codeGeneratorRe]: [
            Pattern,
            RegExp,
            RegExp,
            CodeGeneratorRe,
          ] = args as any
          if (_codeGeneratorRe == null) {
            _codeGeneratorRe = _matchRe as any
            _matchRe = _regionRe as any
            _regionRe = /^.+$/s as any
          }
          if (typeof _file !== 'string') {
            throw new TypeError('Invalid file')
          }
          isRe = true
          fileRe = _file
          regionRe = _regionRe
          matchRe = _matchRe
          codeGeneratorRe = _codeGeneratorRe
        } else {
          const [_patterns, _codeGenerator, _globbyOptions]: [
            Pattern | Pattern[],
            CodeGenerator,
            GlobbyOptions,
          ] = args as any
          if (_patterns == null) {
            throw new TypeError('Invalid patterns')
          }
          patterns = castArray(_patterns)
          codeGenerator = _codeGenerator
          globbyOptions = _globbyOptions
        }
      }
      try {
        eval(`${setParams.name}(${startMatch[2]})`)
        markers.push({
          indent: indent,
          start: start,
          end: end,
          patterns: patterns,
          codeGenerator: codeGenerator,
          globbyOptions: globbyOptions,
          isRe: isRe,
          fileRe: fileRe,
          regionRe: regionRe,
          matchRe: matchRe,
          codeGeneratorRe: codeGeneratorRe,
        })
      } catch (e) {
        onInvalid?.(
          `[SKIP] Invalid patterns or code generator. (${startMatch[0].trim()})`,
        )
      }
      startFrom = end + (endMatch ? endMatch![0].length : 0)
    }

    markers.sort((a, b) => b.start - a.start)

    return markers
  }

  public static normalizePath(path: string): string {
    return path.replace(/[/\\]+/g, '/')
  }

  public static getRelativePath(from: string, to: string): string {
    const relativePath = IndexGenerator.normalizePath(relative(from, to))
    return relativePath.startsWith('../') ? relativePath : `./${relativePath}`
  }
}
