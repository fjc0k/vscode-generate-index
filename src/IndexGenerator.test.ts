import { IndexGenerator } from './IndexGenerator'
import { join } from 'path'
import { readFileSync } from 'fs'

const indexFilePaths = [
  join(__dirname, './__fixtures__/cross/index.txt'),
  join(__dirname, './__fixtures__/scss/index.scss'),
  join(__dirname, './__fixtures__/sort/index.ts'),
  join(__dirname, './__fixtures__/sort/index2.ts'),
  join(__dirname, './__fixtures__/invalid/empty'),
  join(__dirname, './__fixtures__/invalid/error'),
]

test('ok', async () => {
  for (const filePath of indexFilePaths) {
    let fileContent: string = readFileSync(filePath).toString()
    const msgs: string[] = []
    const generator = new IndexGenerator({
      filePath: filePath,
      fileContent: fileContent,
      onWarning: msg => {
        msgs.push(msg)
      },
      onGenerate: ({ code, marker }) => {
        fileContent =
          fileContent.substr(0, marker.start) +
          code +
          fileContent.substr(marker.end, fileContent.length)
      },
    })
    await generator.generate()
    const humanFilePath = filePath.replace(
      join(__dirname, './__fixtures__'),
      '',
    )
    expect(fileContent).toMatchSnapshot(`generated content of ${humanFilePath}`)
    expect(msgs.sort()).toMatchSnapshot(`messages of ${humanFilePath}`)
  }
})
