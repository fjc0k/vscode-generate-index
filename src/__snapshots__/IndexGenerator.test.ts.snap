// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ok: generated content of /all_ts_tsx/index.ts 1`] = `
"// @index(['./*.{ts,tsx}', './*/index.{ts,tsx}'], f => \`export * from '\${f.path.replace(/\\\\/index$/, '')}'\`)
export * from './a'
export * from './b'
export * from './c'
export * from './d'
export * from './j'
// @endindex
"
`;

exports[`ok: generated content of /cross/index.txt 1`] = `
"# @index('../**/*', f => \`\${f.path}\${f.ext}\`, { onlyFiles: true })
../all_ts_tsx/a.ts
../all_ts_tsx/b.ts
../all_ts_tsx/c.tsx
../all_ts_tsx/d/index.ts
../all_ts_tsx/e/hello.ts
../all_ts_tsx/index.ts
../all_ts_tsx/j/index.tsx
../cross_negative/index.txt
../invalid/empty
../invalid/error
../re/index.js
../scripts/index.html
../scripts/jssdk.js
../scripts/polyfill.js
../scss/a.scss
../scss/b.scss
../scss/c.scss
../scss/index.scss
../sort/001.ts
../sort/1.ts
../sort/2.ts
../sort/11.ts
../sort/884.ts
../sort/a.ts
../sort/abc.ts
../sort/bbb.ts
../sort/c.ts
../sort/index.ts
../sort/index2.ts
../sort/国家.ts
../types/icons/arrow.svg
../types/icons/home.svg
../types/icons/pass.svg
../types/icons/picture.svg
../types/icons/user.svg
../types/index.ts
# @endindex

# @index('../**/*', f => \`\${f.path}\${f.ext}\`, { onlyDirectories: true })
../all_ts_tsx
../all_ts_tsx/d
../all_ts_tsx/e
../all_ts_tsx/j
./
../cross_negative
../invalid
../re
../scripts
../scss
../sort
../types
../types/icons
# @endindex

    ! @index('../**/*.scss', (f, _, {isFirst}) => \`\${isFirst ? '****' : ''}\${f.path}\${f.ext}\`)
    ****../scss/a.scss
    ../scss/b.scss
    ../scss/c.scss
    ../scss/index.scss
    ! @endindex
"
`;

exports[`ok: generated content of /cross_negative/index.txt 1`] = `
"# @index(['../*', '!../scss'], f => \`\${f.path}\${f.ext}\`)
../all_ts_tsx
../cross
./
../invalid
../re
../scripts
../sort
../types
# @endindex
"
`;

exports[`ok: generated content of /invalid/empty 1`] = `""`;

exports[`ok: generated content of /invalid/error 1`] = `
"// @index(\\"./*', f => \`export * from '\${f.path}'\`)
"
`;

exports[`ok: generated content of /re/index.js 1`] = `
"// @index('../../../node_modules/fs-extra/lib/index.js', /\\\\.\\\\.\\\\.require\\\\('(.+?)'\\\\)/g, (m, _) => \`\${_.camel(m[1])}\`)
fs
copySync
copy
empty
ensure
json
mkdirs
moveSync
move
output
pathExists
remove
// @endindex

// @index('../../../node_modules/fs-extra/package.json', /(?<=\\"dependencies\\": \\\\{).+?(?=\\\\})/s, /\\"(.+?)\\": \\"(.+?)\\"/g, (m, _, e) => \`\${e.index}. \${m[1]} ==> \${m[2]}\`)
0. at-least-node ==> ^1.0.0
1. graceful-fs ==> ^4.2.0
2. jsonfile ==> ^6.0.1
3. universalify ==> ^2.0.0
// @endindex

  // @index('../../../node_modules/fs-extra/lib/index.js', /\\\\.\\\\.\\\\.require\\\\('(.+?)'\\\\)/g, (m, _) => \`\${_.camel(m[1])}\`)
  fs
  copySync
  copy
  empty
  ensure
  json
  mkdirs
  moveSync
  move
  output
  pathExists
  remove
  // @endindex"
`;

exports[`ok: generated content of /scripts/index.html 1`] = `
"<html>
  <head>
    <!-- @index('./*.js', f => \`<script type=\\"text/javascript\\" src=\\"\${f.path}\${f.ext}\\"></script>\`) -->
    <script type=\\"text/javascript\\" src=\\"./jssdk.js\\"></script>
    <script type=\\"text/javascript\\" src=\\"./polyfill.js\\"></script>
    <!-- @endindex -->
  </head>
  <body>
    ...
  </body>
</html>
"
`;

exports[`ok: generated content of /scss/index.scss 1`] = `
"// @index('./*', f => \`@import '\${f.path}\${f.ext}';\`)
@import './a.scss';
@import './b.scss';
@import './c.scss';
// @endindex
"
`;

exports[`ok: generated content of /sort/index.ts 1`] = `
"// @index('./*.ts', f => \`export * from '\${f.path}'\`)
export * from './001'
export * from './1'
export * from './2'
export * from './11'
export * from './884'
export * from './a'
export * from './abc'
export * from './bbb'
export * from './c'
export * from './index2'
export * from './国家'
// @endindex
"
`;

exports[`ok: generated content of /sort/index2.ts 1`] = `
"// @index('./*.ts', f => \`export * from '\${f.path}'\`)
export * from './001'
export * from './1'
export * from './2'
export * from './11'
export * from './884'
export * from './a'
export * from './abc'
export * from './bbb'
export * from './c'
export * from './index'
export * from './国家'
"
`;

exports[`ok: generated content of /types/index.ts 1`] = `
"export type IconName =
  // @index(['./icons/*.svg'], (f, _, e) => \`'\${f.name}'\${e.isLast ? '' : ' |'}\`)
  'arrow' |
  'home' |
  'pass' |
  'picture' |
  'user'
  // @endindex
"
`;

exports[`ok: messages of /all_ts_tsx/index.ts 1`] = `Array []`;

exports[`ok: messages of /cross/index.txt 1`] = `Array []`;

exports[`ok: messages of /cross_negative/index.txt 1`] = `Array []`;

exports[`ok: messages of /invalid/empty 1`] = `
Array [
  "No @index maker found.",
]
`;

exports[`ok: messages of /invalid/error 1`] = `
Array [
  "No @index maker found.",
  "[SKIP] Invalid patterns or code generator. (// @index(\\"./*', f => \`export * from '\${f.path}'\`))",
]
`;

exports[`ok: messages of /re/index.js 1`] = `Array []`;

exports[`ok: messages of /scripts/index.html 1`] = `Array []`;

exports[`ok: messages of /scss/index.scss 1`] = `Array []`;

exports[`ok: messages of /sort/index.ts 1`] = `Array []`;

exports[`ok: messages of /sort/index2.ts 1`] = `Array []`;

exports[`ok: messages of /types/index.ts 1`] = `Array []`;
